packages/teleterm/src/mainProcess/protocolHandler.ts

```diff
@@ -0,0 +1,88 @@
+import type {
+  protocol as ElectronProtocol,
+  ProtocolRequest,
+  ProtocolResponse,
+} from 'electron';
+import * as path from 'path';
+import fs from 'fs';
+
+type CallbackType = (response: string | ProtocolResponse) => void;
+
+function _disabledHandler(
+  _request: ProtocolRequest,
+  callback: CallbackType
+): void {
+  callback({ error: -3 });
+}
+
+export function installWebHandler({
+  protocol,
+}: {
+  protocol: typeof ElectronProtocol;
+}): void {
+  protocol.interceptFileProtocol('about', _disabledHandler);
+  protocol.interceptFileProtocol('content', _disabledHandler);
+  protocol.interceptFileProtocol('chrome', _disabledHandler);
+  protocol.interceptFileProtocol('cid', _disabledHandler);
+  protocol.interceptFileProtocol('data', _disabledHandler);
+  protocol.interceptFileProtocol('filesystem', _disabledHandler);
+  protocol.interceptFileProtocol('ftp', _disabledHandler);
+  protocol.interceptFileProtocol('gopher', _disabledHandler);
+  protocol.interceptFileProtocol('javascript', _disabledHandler);
+  protocol.interceptFileProtocol('mailto', _disabledHandler);
+}
+
+function eliminateAllAfterCharacter(string, character) {
+  const index = string.indexOf(character);
+  if (index < 0) {
+    return string;
+  }
+
+  return string.slice(0, index);
+}
+
+function urlToPath(targetUrl, { isWin }) {
+  const decoded = decodeURIComponent(targetUrl);
+  const withoutScheme = decoded.slice(isWin ? 8 : 7);
+
+  // remove any extra characters in requsted file path
+  const withoutQuerystring = eliminateAllAfterCharacter(withoutScheme, '?');
+  const withoutHash = eliminateAllAfterCharacter(withoutQuerystring, '#');
+
+  return withoutHash;
+}
+
+// intercept, clean, and validate the requested file path.
+export function interceptFileProtocol({
+  protocol,
+  isWin,
+  userDataPath,
+  installPath,
+  logger,
+}) {
+  protocol.interceptFileProtocol('file', (request, callback) => {
+    const target = path.normalize(urlToPath(request.url, { isWin }));
+    const realPath = fs.existsSync(target) ? fs.realpathSync(target) : target;
+    const properCasing = isWin ? realPath.toLowerCase() : realPath;
+    if (
+      !properCasing.startsWith(
+        isWin ? userDataPath.toLowerCase() : userDataPath
+      ) &&
+      !properCasing.startsWith(isWin ? installPath.toLowerCase() : installPath)
+    ) {
+      logger.error(
+        `Denying request to path '${realPath}' (userDataPath: '${userDataPath}', installPath: '${installPath}'`
+      );
+      return callback();
```

Comment:
Do you happen to know what's the difference in behavior when calling `callback` with an error vs without? I tried to check it by requesting `/etc/passwd` in dev mode but I couldn't get any errors to show up.

Reply:
The errors would only happen in production. In dev, files are still requested over https (from the web server) and only in production do files get requested with the `file://` protocol

Reply:
To answer the first question about callbacks, the error is not needed in the callback. The callback requires an object with a `path` key to be valid, if it doesn't receive one, the request won't work. I will add an error code to them all to make sure it's clear we are NOT returning a successful callback

Reply:
Yeah that error -3 was pretty nifty but I noticed this one was blank.

I was testing this on the packaged built, but when I added a disabled handler under `file` the app wasn't showing any errors. Where should they show up?

I'm mostly worried about future devs running into this and not knowing what's happening.

Reply:
The error was showing for me in `main.log`
example when trying to reference my `/etc/hosts` from the iframe
```
[24-07-22 14:47:16] [Main] error: Denying request to path '/private/etc/hosts' (userDataPath: '/Users/michael/Library/Application Support/Teleport Connect', installPath: '/Users/michael/code/gravitational/webapps/packages/teleterm/build/release/mac-arm64/Teleport Connect.app/Contents/Resources/app.asar'
```

Reply:
I think I understand you're question now. Are you asking "where does the `-3` error show up?"? I'm not sure. I defaulted to returning the 'proper' error code to electron and then using our own error logging in tandem to make sure we see it (in `main.log`). Are there other places we should be seeing logs?

Reply:
Ah, yeah, I just realized that `interceptFileProtocol` will show up in logs. I just realized where my mistake was – I was testing this by adding `protocol.interceptFileProtocol('mailto', _disabledHandler)` but `_disabledHandler` doesn't really have any logging. Could you add it there?